### DISCUSSION TITLE
Are nested `terminating` callbacks a "supported" use case?

### DISCUSSION THREAD

**Discussion by panstromek** [laravel/framework#50713] [Q&A] [answered]
In other words, is registering terminating callaback inside terminating callback supposed to work, or just works accidentaly?

```php
\App::terminating(function () {
    \Log::info('Terminating');

    \App::terminating(function() {
        \Log::info('Post Terminating');
    });
})
````
output:
```
[2024-03-22 09:12:53] local.INFO: Terminating  
[2024-03-22 09:12:53] local.INFO: Post Terminating  
```

This looks like a bit risky thing to do and seems to rely just on the fact that `count($this->terminatingCallbacks)` is not extracted out of the `while` condition in the code that executes these callbacks in [terminate](https://github.com/laravel/framework/blob/08772cf0caaad628570041a58717fe481e315209/src/Illuminate/Foundation/Application.php#L1373) function:

```php
        while ($index < count($this->terminatingCallbacks)) {
            $this->call($this->terminatingCallbacks[$index]);

            $index++;
        }
```

Since just extracting `$count =  count($this->terminatingCallbacks)` could break this use case (and seems like a reasonable optimization someone might want to do), I figured I should first ask if this is intentional or just accidental before using it.

The use case is that I want to log some stats about each request, but the request can add some post processing with `terminating` callback, and I want to include that in the stats, too, so I have to register one more `terminating` callback after all other terminating callbacks.

**Comment by ghost** [comment]
Interesting. If no one uses it like this, I would vote to avoid this case or maybe find a way to restrict it to a maximum of 10 loops or something.

**Comment by rojtjo** [comment, marked as answer]
Yes, this was added intentionally.
See: #39175

**Reply by panstromek** [reply]
Thanks. It'd be good to mention this context somewhere in the comment, I see somebody already tried to change this few times (and docs also don't mention this).
